lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -84,54 +174,91 @@ define((require, exports, module) ->
       getCurrentFormula: ->
         try
           return katex.renderToString(
-            ch.getWholeEquation(ch.curStart, ch.curEnd),
+            ch.getWholeEquation(ch.curRange),
             KATEX_OPTIONS
           )
         catch e
           return e
 
-      initPopover: ->
-        cursorRow = editor.getCursorPosition().row
-        [ch.curStart, ch.curEnd] = ch.getEquationRange(cursorRow)
-        popoverPosition = ch.getPopoverPosition(ch.curEnd)
+      initPopover: -> setTimeout((->
+        popoverPosition = ch.getPopoverPosition(ch.getEquationEnd())
         popoverHandler.show(getFormulaElement(), ch.getCurrentFormula(), popoverPosition)
-        editor.on("change", ch.delayedUpdatePopover)
-        editor.getSession().on("changeScrollTop", ch.updatePosition)
+      ), 0)
+
+      getEquationEnd: ->
+        i = editor.getCursorPosition().row
+        while LatexParsingContext.getContext(editor.getSession(), i) == "equation"
+          i += 1
+        return i
+
+      updatePosition: ->
+        setTimeout((-> popoverHandler.setPosition(getFormulaElement(), ch.getPopoverPosition(ch.getEquationEnd()))), 0)
+
+      updateRange: ->
+        {row: cursorRow, column: cursorColumn} = editor.getCursorPosition()
+        ch.curRange = erh.getEquationRange(cursorRow, cursorColumn)
 
       updatePopover: ->
-        popoverHandler.setContent(getFormulaElement(), ch.getCurrentFormula())
+        if ch.contextPreviewExists
+          popoverHandler.setContent(getFormulaElement(), ch.getCurrentFormula())
 
       updateCallback: ->
         if ch.lastChangeTime?
           curTime = Date.now()
           ch.currentDelayedUpdateId = setTimeout(ch.updateCallback, ch.UPDATE_DELAY - (Date.now() - ch.lastChangeTime))
           ch.lastChangeTime = null
         else
-          ch.updatePopover()
           ch.currentDelayedUpdateId = null
+          curContext = LatexParsingContext.getContext(editor.getSession(), editor.getCursorPosition().row)
```

Comment:
This seem to be used only if `ch.contextPreviewExists == true`, so please move it inside the if